torch/_inductor/fx_passes/ddp_fusion.py

```diff
@@ -0,0 +1,577 @@
+    )
+    move_block_after(incorrect_order_nodes, last_fused_result)
+
+
+def _fuse_allreduce(
+    graph: fx.Graph,
+    comm_blocks: List[CommBlock],
+    node_indices: Dict[fx.Node, int],
+    use_concat: bool,
+) -> CommBlock:
+    """Given a list of allreduce CommBlock, fuse the CommBlocks into one CommBlock."""
+
+    if len(comm_blocks) == 1:
+        return comm_blocks[0]
+
+    # Find the last input node of all the CommBlocks. This node will be served
+    # as the inserting point of the new collective op.
+    last_input_node = comm_blocks[0].inputs[0]
+    last_input_index = -1
+    all_input_nodes = []
+    for comm_block in comm_blocks:
+        input_node = comm_block.inputs[0]
+        all_input_nodes.append(input_node)
+        index = node_indices[input_node]
+        if index >= last_input_index:
+            assert index != last_input_index
+            last_input_node = input_node
+            last_input_index = index
+
+    if use_concat:
+        fused_comm_block = _fuse_allreduce_by_concat(
+            graph, last_input_node, all_input_nodes, comm_blocks[-1]
+        )
+    else:
+        fused_comm_block = _fuse_with_coalesced_op(
+            graph, last_input_node, all_input_nodes, comm_blocks[-1]
+        )
+
+    _scatter_fused_allreduce_waits(
+        graph, fused_comm_block, comm_blocks, node_indices, split_and_reshape=use_concat
+    )
+
+    for comm_block in comm_blocks:
+        for wait in comm_block.wait_nodes:
+            graph.erase_node(wait)
+        graph.erase_node(comm_block.comm_node)
+    graph.eliminate_dead_code()
+
+    return fused_comm_block
+
+
+def _bucket_size_fusion(
+    graph: fx.Graph, comm_blocks: List[CommBlock], bucket_size_mb: int
+) -> Generator[List[CommBlock], None, None]:
+    MB = 1024**2
+    bucket_size = 1 * MB
+    bucket_cap_size = bucket_size_mb * MB
+    curr_size = 0
+    curr_blocks = []
+
+    count = 0
+    fuse_count = 0
+    for i, block in enumerate(comm_blocks):
+        curr_blocks.append(block)
+        itemsize = block.comm_node.meta["tensor_meta"].dtype.itemsize
+        curr_size += cast(torch.Size, block.shape).numel() * itemsize
+        count += 1
+        if curr_size < bucket_size and i != len(comm_blocks) - 1:
+            continue
+
+        fuse_count += 1
+        if torch.distributed.get_rank() == 0:
+            logger.info(
+                "DDP bucketing: block%d, count=%d, curr_size=%d, bucket_size=%d",
+                fuse_count,
+                count,
+                curr_size,
+                bucket_size,
+            )
+
+        # Set the debug counters
+        counters["inductor"]["ddp_buckets"] = fuse_count
+        yield curr_blocks
+
+        bucket_size = bucket_cap_size
+        curr_blocks = []
+        curr_size = 0
+        count = 0
+
+
+def _fuse_ddp_communication(
+    graph: fx.Graph, algorithm_fn: Callable[..., Any], fusion_fn: Callable[..., Any]
+) -> None:
+    def ddp_reducer_filter(block: CommBlock) -> bool:
```

Comment:
i wonder if using the standard fx pattern match tooling would have been just as good here?  (and also for other 'comm_block' use cases? not sure how different it would be, so i'm not too picky here)

Reply:
Unlike most other pattern match passes, communication fusion needs to get all the matched patterns, CommBlocks in this PR, first before making any replacement or decision. We cannot make any decision based on only one matched pattern. For example, even if two CommBlocks have 2MB tensors in total, we may still want to fuse them if these two are the last two CommBlocks. 

So we need the following steps:
1. Get all the matched CommBlocks.
2. Scan all the CommBlocks to make the grouping decision.
3. Apply the replacement to multiple CommBlocks for each bucket.

I think we can extend the pattern matcher to support the steps, at least some of the steps. However, what I don't know is that if we are going to move the communication fusion to Inductor IR level, is it still worth to invest on building the infra in FX IR level?

As for the UX, I admit it is not ideal to expose these options. Maybe we can just change them to private. We need these options at this moment because any change in NCCL or Inductor may affect the performance. Right now, there is no good way for us to auto detect which one is going to be the best.